### PR TITLE
support path rules

### PR DESCRIPTION
**Pull request by nightlyone**
by allowing a placeholder.

explicit paths still have precedence.

This is useful in migration scenarios and when you don't want to deploy
just because someone added a repo to you organisation account.

### CODE REVIEW

**Comment by nightlyone** [comment]
Hi @rakyll 

please take a look and give early feedback on the overall approach and what else you see.

Once it looks good, I will update the readme to reflect the new possibilities.

Many thanks :slightly_smiling_face:

**Comment by nightlyone** [comment]
gentle ping @rakyll 

If you are in a hurry: To get a :+1: or :-1: on the basic idea without you reviewing everything,
having early feedback on https://github.com/GoogleCloudPlatform/govanityurls/pull/12/files#diff-c6c82e356db0ffdc1ce631b75c55e263R100 would be most useful to me, since those 25 lines it describe the intended new behavior best.

File: handler.go [hunk 1]
```diff
@@ -37,20 +38,105 @@ type pathConfig struct {
...
+	i = strings.Index(s, "}")
+	if i < 0 {
+		 return "", "", "", fmt.Errorf("placeholder not terminated in %q", match)
 	}
-	h := &handler{host: parsed.Host}
-	for path, e := range parsed.Paths {
+	if i == 0 {
+		 return "", "", "", fmt.Errorf("placeholder is empty in %q", match)
+	}
+	placeholder, s = s[:i], s[i+1:]
```
Comment (rakyll) [inline handler.go]: placeholder, suffix = s[:i], s[i+1:]

File: handler.go [hunk 2]
```diff
@@ -37,20 +38,105 @@ type pathConfig struct {
...
+	suffix = s
+	if strings.ContainsAny(suffix, "{}") {
```
Comment (rakyll) [inline handler.go]: This is not going to match if there is anything between curly braces.

File: handler.go [hunk 3]
```diff
@@ -26,8 +26,9 @@ import (
 )
 
 type handler struct {
-	host  string
-	paths pathConfigSet
+	host      string
+	paths     pathConfigSet
```
Comment (rakyll) [inline handler.go]: why do we need both parsed paths and parsedPathRules? Can't we merge them into one type?

File: handler.go [hunk 4]
```diff
@@ -37,20 +38,105 @@ type pathConfig struct {
...
 
-func newHandler(config []byte) (*handler, error) {
-	var parsed struct {
-		Host  string `yaml:"host,omitempty"`
-		Paths map[string]struct {
-			Repo    string `yaml:"repo,omitempty"`
-			Display string `yaml:"display,omitempty"`
-			VCS     string `yaml:"vcs,omitempty"`
-		} `yaml:"paths,omitempty"`
+type pathConfigRule struct {
```
Comment (rakyll) [inline handler.go]: These need to go into the pathConfig instead of introducing a new type.

File: handler.go [hunk 5]
```diff
@@ -137,9 +254,13 @@ func (h *handler) Host(r *http.Request) string {
 var indexTmpl = template.Must(template.New("index").Parse(`<!DOCTYPE html>
 <html>
 <h1>{{.Host}}</h1>
+
 <ul>
 {{range .Handlers}}<li><a href="https://godoc.org/{{.}}">{{.}}</a></li>{{end}}
 </ul>
+<ul>
+{{range .GenericRules}}<li>{{.Match}} will be clone repository {{.Subst}}></li>{{end}}
```
Comment (rakyll) [inline handler.go]: Just add the rule to the list above without a link and a placeholder.

**Comment by nightlyone** [comment]
@rakyll fixed missing subpath feature and added a testcase for that. Addressed the smaller code issues.

About using only one struct to do both, I need a bit more time to think about how to do that. Will come back to you tomorrow with more ideas.

**Comment by rakyll** [comment]
@nightlyone, thanks.

Also cc'ing @zombiezen for the review.

File: handler.go (hunk 3 above)
Comment (zombiezen) [reply handler.go]: :+1:

**Comment by nightlyone** [comment]
Got lost here so closing to get it off everyones table.

**Pull request closed without merging**